src/security/builtin_plugins/cryptographic/src/crypto_transform.c

```diff
@@ -1630,93 +1271,79 @@ static DDS_Security_boolean encode_rtps_message_encrypt (
-  size += sizeof(struct submsg_header) + RTPS_HEADER_SIZE; /* INFO_SRC */
+  size += sizeof(info_src); /* INFO_SRC */
 
   if (is_encryption_required(session->master_key_material->transformation_kind))
-    size += sizeof(struct submsg_header) + sizeof(uint32_t);
+    size += sizeof(struct crypto_body);
 
   /* allocate a buffer to store the encoded message */
-  data._buffer = ddsrt_malloc(size);
-  data._length = 0;
-  data._maximum = (uint32_t)size;
-
-  /* Add RTPS header */
-  memcpy(data._buffer, plain_rtps_message->_buffer, RTPS_HEADER_SIZE);
-  data._length += RTPS_HEADER_SIZE;
+  crypto_buffer_init(&buffer, size);
 
-  /* Set the SEC_PREFIX and associated CryptoHeader */
-  prefix = add_submessage(&data, SMID_SRTPS_PREFIX_KIND, flags, sizeof(struct crypto_header));
-  header = (struct crypto_header *)(prefix + 1);
-  contents = (unsigned char *)(header + 1);
-
-  /* update sessionKey when needed */
-  if (!crypto_session_key_material_update(session, secure_body_plain_size, ex))
-    goto enc_rtps_fail_data;
+  plain_data[0].base = (unsigned char *)&info_src;
+  plain_data[0].length = sizeof(info_src);
+  plain_data[1].base = (unsigned char *)(rtps_header + 1);
+  plain_data[1].length = plain_rtps_message->_length - RTPS_MESSAGE_HEADER_SIZE;
 
   /* increment init_vector_suffix */
   session->init_vector_suffix++;
 
   transform_kind = session->master_key_material->transformation_kind;
   transform_id = session->master_key_material->sender_key_id;
-  set_crypto_header(header, transform_kind, transform_id, session->id, session->init_vector_suffix);
+
+  /* copy the rtps header to the encryption buffer */
+  (void)add_rtps_header(&buffer, rtps_header);
+
+  header = add_crypto_header(&buffer, SMID_SRTPS_PREFIX, transform_kind, transform_id, session->id, session->init_vector_suffix);
 
   if (is_encryption_required(transform_kind))
   {
-    struct encrypted_data *encrypted;
-    size_t submsg_len = secure_body_plain_size + sizeof (uint32_t);
-    if (submsg_len > UINT16_MAX)
-      goto enc_rtps_fail_data;
+    struct crypto_body *body = add_crypto_body(&buffer, secure_body_plain_size);
 
-    /* add SEC_BODY submessage */
-    body = add_submessage(&data, SMID_SEC_BODY_KIND, flags, submsg_len);
-    encrypted = (struct encrypted_data *)(body + 1);
-    contents = encrypted->data;
+    encrypted_data.base = body->content.data;
+    encrypted_data.length = secure_body_plain_size;
 
     /* encrypt message */
-    /* FIXME: improve performance by not copying plain_rtps_message to a new buffer (crypto_cipher_encrypt_data should allow encrypting parts of a message) */
-    if (!crypto_cipher_encrypt_data(&session->key, session->key_size, header->session_id, secure_body_plain._buffer, secure_body_plain_size, NULL, 0, contents, &payload_len, &hmac, ex))
+    if (!crypto_cipher_encrypt_data(&session->key, session->key_size, &header->prefix.iv, 2, plain_data, &encrypted_data, &hmac, ex))
```

Comment:
There's something to be said for doing `sizeof(plain_data)/sizeof(plain_data[0])`, it didn't take me long to figure out what that `2` means and why it is correct, but I do like how a `sizeof` here would be somewhat self-documenting. (Same holds for some similar places.)